vcf.c

```diff
@@ -6137,3 +6139,246 @@ int bcf_format_gt_v2(const bcf_hdr_t *hdr, bcf_fmt_t *fmt, int isample, kstring_
     }
     return e == 0 ? 0 : -1;
 }
+
+/**
+ *  get_rlen - calculates and returns rlen value
+ *  @param h - bcf header
+ *  @param v - bcf data
+ *  Returns rlen calculated on success and -1 on failure.
+ *  rlen calculation is dependent on vcf version and a few other field data.
+ *  When bcf decoded data is available, refers it. When not available, retrieves
+ *  required field data by seeking on the data stream.
+ *  pos, version has to be set appropriately before any info/format field
+ *  update to have proper rlen calculation.
+ *  TODO version change has to trigger rlen recalculation on whole data
+ */
+static int64_t get_rlen(const bcf_hdr_t *h, bcf1_t *v)
+{
+    uint8_t *f = (uint8_t*)v->shared.s, *t = NULL, *e = (uint8_t*)v->shared.s + v->shared.l;
+    int size, type, id, lenid, endid, svlenid, i, bad;
+    char **alleles = calloc(1, sizeof(char*) * v->n_allele);
+    bcf_info_t *endinfo = NULL, *svleninfo = NULL;
+    bcf_fmt_t *lenfmt = NULL;
+    //free each allele, end info, svlen info, len format or not
+    int free_als = 0, free_end = 0, free_svlen = 0, free_len = 0;
+
+    //pos from info END, fmt LEN, info SVLEN
+    hts_pos_t end = 0, end_fmtlen = 0, end_svlen = 0;
+    int64_t len_ref = 0, len_end = 0, len_fmtlen = 0, len_svlen = 0, len = 0;
+    int64_t rlen = -1, tmp;
+    int version = bcf_get_version(h, NULL);
+
+    lenid = bcf_hdr_id2int(h, BCF_DT_ID, "LEN");
+    svlenid = bcf_hdr_id2int(h, BCF_DT_ID, "SVLEN");
+    endid = bcf_hdr_id2int(h, BCF_DT_ID, "END");
+
+    //use decoded data where ever available and where not, get from stream
+    if (v->unpacked & BCF_UN_STR || v->d.shared_dirty & BCF1_DIRTY_ALS) {
+        for (i = 0; i < v->n_allele; ++i) {
+            alleles[i] = v->d.allele[i];
+        }
+        f += v->unpack_size[0] + v->unpack_size[1];
+    } else if (f < e) {
+        //skip ID
+        size = bcf_dec_size(f, &f, &type);
+        f += size << bcf_type_shift[type];
+        // REF, ALT
+        for (i = 0; i < v->n_allele; ++i) {
+            size = bcf_dec_size(f, &f, &type);
+            if (!(alleles[i] = malloc(sizeof(char) * (size + 1))))
+                goto fail;
+            memcpy(alleles[i], f, size);
+            alleles[i][size] = '\0';
+            f += size << bcf_type_shift[type];
+        }
+        free_als = 1;   //free at end
+    }
+    // FILTER
+    if (v->unpacked & BCF_UN_FLT) {
+        f += v->unpack_size[2];
+    } else if (f < e) {
+        size = bcf_dec_size(f, &f, &type);
+        f += size << bcf_type_shift[type];
+    }
+    // INFO
+    if (v->unpacked & BCF_UN_INFO || v->d.shared_dirty & BCF1_DIRTY_INF) {
+        endinfo = bcf_get_info(h, v, "END");
+        svleninfo = version >= VCF44 ? bcf_get_info(h, v, "SVLEN") : NULL;
+    } else if (f < e) {
+        for (i = 0; i < v->n_info; ++i) {
+            id = bcf_dec_typed_int1(f, &t);
+            if (id == endid) {  //END
+                if (!(endinfo = malloc(sizeof(bcf_info_t))))
```

Comment:
You only ever need one of these, so could it be allocated on the stack, along with a flag to say that it's in use?

Similarly for `svleninfo` and `lenfmt`.